### PR TITLE
 KAFKA-6054: Fix upgrade path from Kafka Streams v0.10.0

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
This is the port of #4746 (`0.10.1`), #4758 (`0.10.2`), #4761 (`0.11.0`), and #4768 (`1.0`) to branch `1.1`.

> Note: `1.0` was the first branch that did contain upgrade system tests already -- I just added the new once -- we can consider to consolidate them in this PR already. LMKWYT.

If we decide to cleanup tests in 1.0, those changes need to be added to this branch, too. System tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1605/

\cc @bbejeck @vvcephei

**Comment by mjsax** [comment]
System tests passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1605/

File: tests/kafkatest/tests/streams/streams_upgrade_test.py [hunk 1]
```diff
@@ -165,3 +167,253 @@ def test_upgrade_brokers(self, from_version, to_version):
...
+    @parametrize(old_version=str(LATEST_0_10_1), new_version=str(LATEST_0_11_0))
+    @parametrize(old_version=str(LATEST_0_10_1), new_version=str(LATEST_1_0))
+    @parametrize(old_version=str(LATEST_0_10_1), new_version=str(DEV_VERSION))
+    @parametrize(old_version=str(LATEST_0_10_2), new_version=str(LATEST_0_11_0))
+    @parametrize(old_version=str(LATEST_0_10_2), new_version=str(LATEST_1_0))
+    @parametrize(old_version=str(LATEST_0_10_2), new_version=str(DEV_VERSION))
+    @parametrize(old_version=str(LATEST_0_11_0), new_version=str(LATEST_1_0))
+    @parametrize(old_version=str(LATEST_0_11_0), new_version=str(DEV_VERSION))
+    @parametrize(old_version=str(LATEST_1_0), new_version=str(DEV_VERSION))
+    def test_simple_upgrade(self, old_version, new_version):
```
Comment (guozhangwang) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py]: Similar to my other PR's comment, I'd suggest we update this test to get rid of using StreamsSmokeTestDriverService and use StreamsUpgradeTestJobRunnerService.

Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: Can you elaborate? `StreamsSmokeTestDriverService` is used the generate the input data in this test and there is no alternative implementation, and `StreamsUpgradeTestJobRunnerService` is used already.

File: docs/streams/upgrade-guide.html [hunk 2]
```diff
@@ -34,13 +34,7 @@ <h1>Upgrade Guide and API Changes</h1>
...
 
     <p>
-        If you want to upgrade from 1.1.x to 1.2.0 and you have customized window store implementations on the <code>ReadOnlyWindowStore</code> interface
-        you'd need to update your code to incorporate the newly added public APIs; otherwise you don't need to make any code changes.
-        See <a href="#streams_api_changes_120">below</a> for a complete list of 1.2.0 API and semantic changes that allow you to advance your application and/or simplify your code base.
-    </p>
-
-    <p>
-        If you want to upgrade from 1.0.x to 1.1.0 and you have customized window store implementations on the <code>ReadOnlyWindowStore</code> interface
+        If you want to upgrade from 1.0.x to 1.1.1 and you have customized window store implementations on the <code>ReadOnlyWindowStore</code> interface
```
Comment (vvcephei) [inline docs/streams/upgrade-guide.html]: Should these target versions be like `1.1.x`?

(I'll not repeat this question on every occurrence)

**Review by vvcephei — APPROVED** [review]
This PR LGTM.

I do agree that we could consider consolidating system tests, but I also think we can call the scope on this PR big enough already. I think it's a better idea to make a separate task to step back and look at our whole testing strategy to try and globally optimize it.

**Comment by mjsax** [comment]
Rebased, cherry-picked missing changes from #4768, and address review comments. Triggered upgrade system test: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1648/

Call for review.

**Comment by guozhangwang** [comment]
The system test 1648 failed, is it related?

**Comment by mjsax** [comment]
I think so. Need to investigate... (maybe I introduced an issue rebasing or cherry-picking)

**Review by vvcephei — APPROVED** [review]
This looks right to me.

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by mjsax** [comment]
Retriggered system tests: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1651/

1.1.0 release broke the tests -- needed to upgrade version for tests to 1.1.1-SNAPSHOT

**Comment by mjsax** [comment]
The failing tests (`test_upgrade_downgrade_brokers`), all crash with
```
Exception in thread "main" java.lang.UnsupportedClassVersionError: kafka/Kafka : Unsupported major.minor version 52.0
```
on broker startup affecting version 1.0.1 only. We disabled this test in `trunk`. Should we disable for this PR -- issue seems to be unrelated?

**Comment by vvcephei** [comment]
I think this error is saying that Kafka was compiled targeting JDK8 but you're attempting to run it with JDK7. If that sounds like the kind of thing we control in the system tests, I think we should fix it rather than disabling the test.

**Comment by vvcephei** [comment]
@mjsax I forgot to @-mention you about this^^

**Comment by mjsax** [comment]
Retriggered as https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1653/

**Comment by mjsax** [comment]
System test passed. Call for review.

@vvcephei Scala version was the issue -- I downgraded to use 2.11. System test are executed using JDK7 in branch builder.

**Review by guozhangwang — APPROVED** [review]

**Pull request merged**